File: packages/plplot/examples/c++/x16.cc
//---------------------------------------------------------------------------//
// $Id: x16.cc 3186 2006-02-15 18:17:33Z slbrow $
//---------------------------------------------------------------------------//
//
//---------------------------------------------------------------------------//
// Copyright (C) 2004  Andrew Ross <andrewr@coriolis.greenend.org.uk>
// Copyright (C) 2004  Alan W. Irwin
//
// This file is part of PLplot.
//
// PLplot is free software; you can redistribute it and/or modify
// it under the terms of the GNU Library General Public License as published by
// the Free Software Foundation; version 2 of the License.
//
// PLplot is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Library General Public License for more details.
//
// You should have received a copy of the GNU Library General Public License
// along with PLplot; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
//---------------------------------------------------------------------------//
//
//---------------------------------------------------------------------------//
// Implementation of PLplot example 16 in C++.
//---------------------------------------------------------------------------//

#include "plstream.h"

#include <iostream>
#include <cmath>

#ifdef USE_NAMESPACE
using namespace std;
#endif

class x16 {

public:
  x16(int, char **);

private:
  // Class data
  plstream *pls;

  // number of shade levels, x data points, y data points,
  // and perimeter points.
  static int ns;
  static int nx;
  static int ny;
  static int exclude;
  const static int PERIMETERPTS;

  // calculated constants and array that depends on them
  PLFLT xspa;
  PLFLT ypsa;
public:
  static PLFLT tr[];

  static PLOptionTable options[];
  static char *notes[];

};

int x16::ns = 20;
int x16::nx = 35;
int x16::ny = 46;
int x16::exclude = 0;
const int x16::PERIMETERPTS = 100;

PLFLT x16::tr[] = {2./(nx-1), 0.0, -1.0, 0.0, 2./(ny-1), -1.0};


/* Options data structure definition. */

PLOptionTable x16::options[] = {
{
    "exclude",                  /* Turns on page showing exclusion */
    NULL,
    NULL,
    &exclude,
    PL_OPT_BOOL,
    "-exclude",
    "Plot the \"exclusion\" page." },
{
    "ns",                       /* Number of shade levels */
    NULL,
    NULL,
    &ns,
    PL_OPT_INT,
    "-ns levels",
    "Sets number of shade levels" },
{
    "nx",                       /* Number of data points in x */
    NULL,
    NULL,
    &nx,
    PL_OPT_INT,
    "-nx xpts",
    "Sets number of data points in x" },
{
    "ny",                       /* Number of data points in y */
    NULL,
    NULL,
    &ny,
    PL_OPT_INT,
    "-ny ypts",
    "Sets number of data points in y" },
{
    NULL,                       /* option */
    NULL,                       /* handler */
    NULL,                       /* client data */
    NULL,                       /* address of variable to set */
    0,                          /* mode flag */
    NULL,                       /* short syntax */
    NULL }                      /* long syntax */
};

char *x16::notes[] = {
"To get smoother color variation, increase ns, nx, and ny.  To get faster",
"response (especially on a serial link), decrease them.  A decent but quick",
"test results from ns around 5 and nx, ny around 25.",
NULL};

/* Transformation function */

static void mypltr(PLFLT x, PLFLT y, PLFLT *tx, PLFLT *ty, void *pltr_data) {
  *tx = x16::tr[0] * x + x16::tr[1] * y + x16::tr[2];
  *ty = x16::tr[3] * x + x16::tr[4] * y + x16::tr[5];
}

PLINT zdefined (PLFLT x, PLFLT y) {
  PLFLT z = sqrt(x * x + y * y);

  return z < 0.4 || z > 0.6;
}

x16::x16( int argc, char ** argv ) {

  int i, j;
  PLFLT x, y, argx, argy, distort, r, t, zmin, zmax;
  PLFLT *px = new PLFLT[PERIMETERPTS];
  PLFLT *py = new PLFLT[PERIMETERPTS];
  PLFLT *xg1 = new PLFLT[nx];
  PLFLT *yg1 = new PLFLT[ny];
  PLFLT **z;
  PLFLT **w;
  PLcGrid cgrid1;
  PLcGrid2 cgrid2;
  PLFLT *clevel = new PLFLT[ns];
  PLFLT *shedge = new PLFLT[ns+1];
  const int fill_width = 2, cont_color = 0, cont_width = 0;


  // plplot initialization

  pls = new plstream();

  // Parse and process command line arguments.
  pls->MergeOpts(options, "x16c options", notes);
  pls->parseopts( &argc, argv, PL_PARSE_FULL );

  // Reduce colors in cmap 0 so that cmap 1 is useful on a
  //16-color display
  pls->scmap0n(3);


  // Initialize PLplot.
  pls->init();

  pls->Alloc2dGrid(&z,nx,ny);
  pls->Alloc2dGrid(&w,nx,ny);

  // calculated constants and array that depends on them
  PLFLT xspa =  2./(nx-1);
  PLFLT yspa =  2./(ny-1);
  PLFLT tr[] = {xspa, 0.0, -1.0, 0.0, yspa, -1.0};



  // Set up data array

  for (i = 0; i < nx; i++) {
    x = (PLFLT) (i - (nx / 2)) / (PLFLT) (nx / 2);
    for (j = 0; j < ny; j++) {
      y = (PLFLT) (j - (ny / 2)) / (PLFLT) (ny / 2) - 1.0;

      z[i][j] = - sin(7.*x) * cos(7.*y) + x*x - y*y;
      w[i][j] = - cos(7.*x) * sin(7.*y) + 2 * x * y;
    }
  }

  pls->MinMax2dGrid(z, nx, ny, &zmax, &zmin );

  for (i = 0; i < ns; i++)
    clevel[i] = zmin + (zmax - zmin) * (i + 0.5) / (PLFLT) ns;

  for (i = 0; i < ns+1; i++)
    shedge[i] = zmin + (zmax - zmin) * (PLFLT) i / (PLFLT) ns;

/* Set up coordinate grids */

  cgrid1.xg = xg1;
  cgrid1.yg = yg1;
  cgrid1.nx = nx;
  cgrid1.ny = ny;

  pls->Alloc2dGrid(&cgrid2.xg, nx, ny);
  pls->Alloc2dGrid(&cgrid2.yg, nx, ny);
  cgrid2.nx = nx;
  cgrid2.ny = ny;

  for (i = 0; i < nx; i++) {
    for (j = 0; j < ny; j++) {
      mypltr((PLFLT) i, (PLFLT) j, &x, &y, NULL);

      argx = x * M_PI/2;
      argy = y * M_PI/2;
      distort = 0.4;

      cgrid1.xg[i] = x + distort * cos(argx);
      cgrid1.yg[j] = y - distort * cos(argy);

      cgrid2.xg[i][j] = x + distort * cos(argx) * cos(argy);
      cgrid2.yg[i][j] = y - distort * cos(argx) * cos(argy);
    }
  }

  // Plot using identity transform

  pls->adv(0);
  pls->vpor(0.1, 0.9, 0.1, 0.9);
  pls->wind(-1.0, 1.0, -1.0, 1.0);

  pls->psty(0);

  pls->shades(z, nx, ny, NULL, -1., 1., -1., 1.,
	       shedge, ns+1, fill_width,
	       cont_color, cont_width,
	       plstream::fill, 1, NULL, NULL);

  pls->col0(1);
  pls->box("bcnst", 0.0, 0, "bcnstv", 0.0, 0);
  pls->col0(2);
  pls->lab("distance", "altitude", "Bogon density");

  // Plot using 1d coordinate transform

  pls->adv(0);
  pls->vpor(0.1, 0.9, 0.1, 0.9);
  pls->wind(-1.0, 1.0, -1.0, 1.0);

  pls->psty(0);

  pls->shades(z, nx, ny, NULL, -1., 1., -1., 1.,
	       shedge, ns+1, fill_width,
	       cont_color, cont_width,
	       plstream::fill, 1, plstream::tr1, (void *) &cgrid1);

  pls->col0(1);
  pls->box("bcnst", 0.0, 0, "bcnstv", 0.0, 0);
  pls->col0(2);

  pls->lab("distance", "altitude", "Bogon density");

  // Plot using 2d coordinate transform

  pls->adv(0);
  pls->vpor(0.1, 0.9, 0.1, 0.9);
  pls->wind(-1.0, 1.0, -1.0, 1.0);

  pls->psty(0);

  pls->shades(z, nx, ny, NULL, -1., 1., -1., 1.,
	       shedge, ns+1, fill_width,
	       cont_color, cont_width,
	       plstream::fill, 0, plstream::tr2, (void *) &cgrid2);

  pls->col0(1);
  pls->box("bcnst", 0.0, 0, "bcnstv", 0.0, 0);
  pls->col0(2);
  pls->cont(w, nx, ny, 1, nx, 1, ny, clevel, ns, plstream::tr2, (void *) &cgrid2);

  pls->lab("distance", "altitude", "Bogon density, with streamlines");

  // Plot using 2d coordinate transform

  pls->adv(0);
  pls->vpor(0.1, 0.9, 0.1, 0.9);
  pls->wind(-1.0, 1.0, -1.0, 1.0);

  pls->psty(0);

  pls->shades(z, nx, ny, NULL, -1., 1., -1., 1.,
	       shedge, ns+1, fill_width,
	       2, 3,
	       plstream::fill, 0, plstream::tr2, (void *) &cgrid2);

  pls->col0(1);
  pls->box("bcnst", 0.0, 0, "bcnstv", 0.0, 0);
  pls->col0(2);

  pls->lab("distance", "altitude", "Bogon density");

  // Note this exclusion API will probably change
  // Plot using 2d coordinate transform and exclusion

  if(exclude) {
    pls->adv(0);
    pls->vpor(0.1, 0.9, 0.1, 0.9);
    pls->wind(-1.0, 1.0, -1.0, 1.0);

    pls->psty(0);

    pls->shades(z, nx, ny, zdefined, -1., 1., -1., 1.,
             shedge, ns+1, fill_width,
             cont_color, cont_width,
             plstream::fill, 0, plstream::tr2, (void *) &cgrid2);

    pls->col0(1);
    pls->box("bcnst", 0.0, 0, "bcnstv", 0.0, 0);

    pls->lab("distance", "altitude", "Bogon density with exclusion");

    }

  //Example with polar coordinates.

  pls->adv(0);
  pls->vpor( .1, .9, .1, .9 );
  pls->wind( -1., 1., -1., 1. );

  pls->psty(0);

  // Build new coordinate matrices.

  for (i = 0; i < nx; i++) {
    r = ((PLFLT) i)/ (nx-1);
    for (j = 0; j < ny; j++) {
      t = (2.*M_PI/(ny-1.))*j;
      cgrid2.xg[i][j] = r*cos(t);
      cgrid2.yg[i][j] = r*sin(t);
      z[i][j] = exp(-r*r)*cos(5.*M_PI*r)*cos(5.*t);
    }
  }

  //Need a new shedge to go along with the new data set.

  pls->MinMax2dGrid(z, nx, ny, &zmax, &zmin );

  for (i = 0; i < ns+1; i++)
    shedge[i] = zmin + (zmax - zmin) * (PLFLT) i / (PLFLT) ns;

  pls->shades(z, nx, ny, NULL, -1., 1., -1., 1.,
	       shedge, ns+1, fill_width,
	       cont_color, cont_width,
	       plstream::fill, 0, plstream::tr2, (void *) &cgrid2);

  // Now we can draw the perimeter.  (If do before, shade stuff may overlap.)
  for (i = 0; i < PERIMETERPTS; i++) {
    t = (2.*M_PI/(PERIMETERPTS-1))*(PLFLT)i;
    px[i] = cos(t);
    py[i] = sin(t);
  }
  pls->col0(1);
  pls->line(PERIMETERPTS, px, py);

  // And label the plot.

  pls->col0(2);
  pls->lab( "", "",  "Tokamak Bogon Instability" );

  // Clean up

  // pls->end();

  pls->Free2dGrid(z,nx,ny);
  pls->Free2dGrid(w,nx,ny);
  pls->Free2dGrid(cgrid2.xg, nx, ny);
  pls->Free2dGrid(cgrid2.yg, nx, ny);

  delete[] px;
  delete[] py;
  delete[] xg1;
  delete[] yg1;
  delete[] clevel;
  delete[] shedge; 
  delete pls;

}

int main( int argc, char ** argv ) {
  x16 *x = new x16( argc, argv );

  delete x;
}


//---------------------------------------------------------------------------//
//                              End of x16.cc
//---------------------------------------------------------------------------//

File: packages/plplot/examples/java/x16.java
//---------------------------------------------------------------------------//
// $Id: x16.java 3186 2006-02-15 18:17:33Z slbrow $
//---------------------------------------------------------------------------//

//---------------------------------------------------------------------------//
// Copyright (C) 2001  Geoffrey Furnish
// Copyright (C) 2002  Alan W. Irwin
// Copyright (C) 2004  Andrew Ross
//
// This file is part of PLplot.
//
// PLplot is free software; you can redistribute it and/or modify
// it under the terms of the GNU Library General Public License as published by
// the Free Software Foundation; version 2 of the License.
//
// PLplot is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Library General Public License for more details.
//
// You should have received a copy of the GNU Library General Public License
// along with PLplot; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
//---------------------------------------------------------------------------//

//---------------------------------------------------------------------------//
// Implementation of PLplot example 16 in Java.
//---------------------------------------------------------------------------//

package plplot.examples;

import plplot.core.*;

import java.lang.Math;

class x16 {

   // number of shade levels, x data points, y data points, and perimeter points.
   static final int NSHADES = 20;
   static final int XPTS = 35;
   static final int YPTS = 46;
   static final int PERIMETERPTS = 100;

   // calculated constants and array that depends on them
   static final double XSPA =  2./(XPTS-1);
   static final double YSPA =  2./(YPTS-1);
   final double tr[] = {XSPA, 0.0, -1.0, 0.0, YSPA, -1.0};

   PLStream pls = new PLStream();

   double fmin, fmax;

   // Compute min and max value of a 2-d array.

   void f2mnmx( double[][] f, int nx, int ny )
     {
        fmax = f[0][0];
        fmin = fmax;
	
        for( int i=0; i < nx; i++ )
	  for( int j=0; j < ny; j++ ) {
	     if (f[i][j] < fmin) fmin = f[i][j];
	     if (f[i][j] > fmax) fmax = f[i][j];
	  }
     }

   // Does a variety of shade plots with continuous (cmap1) colours

   public static void main( String[] args ) 
     {
        x16 x = new x16( args );
     }

   public x16( String[] args )
     {
	int i, j;
	double x, y, argx, argy, distort, r, t, zmin, zmax;
	double[] px = new double[PERIMETERPTS];
	double[] py = new double[PERIMETERPTS];
	double[][] xg0 = new double[XPTS][YPTS];
	double[][] yg0 = new double[XPTS][YPTS];
	double[][] xg1 = new double[XPTS][YPTS];
	double[][] yg1 = new double[XPTS][YPTS];
	double[][] z = new double[XPTS][YPTS];
	double[][] w = new double[XPTS][YPTS];
	double[][] xg2 = new double[XPTS][YPTS];
	double[][] yg2 = new double[XPTS][YPTS];
	double[] clevel = new double[NSHADES];
	double[] shedge = new double[NSHADES+1];
	final int fill_width = 2, cont_color = 0, cont_width = 0;

	// Parse and process command line arguments.
	pls.parseopts( args, pls.PL_PARSE_FULL | pls.PL_PARSE_NOPROGRAM );

	// Reduce colors in cmap 0 so that cmap 1 is useful on a 
	//16-color display
	pls.scmap0n(3);

	//Initialize plplot
         
          pls.init();
	// Set up data array

	for (i = 0; i < XPTS; i++) {
	   x = (double) (i - (XPTS / 2)) / (double) (XPTS / 2);
	   for (j = 0; j < YPTS; j++) {
	      y = (double) (j - (YPTS / 2)) / (double) (YPTS / 2) - 1.0;

	      z[i][j] = - Math.sin(7.*x) * Math.cos(7.*y) + x*x - y*y;
	      w[i][j] = - Math.cos(7.*x) * Math.sin(7.*y) + 2 * x * y;
	   }
	}

	f2mnmx( z, XPTS, YPTS );
	zmin = fmin;
	zmax = fmax;
	
	for (i = 0; i < NSHADES; i++)
	  clevel[i] = zmin + (zmax - zmin) * (i + 0.5) / (double) NSHADES;

	for (i = 0; i < NSHADES+1; i++)
	  shedge[i] = zmin + (zmax - zmin) * (double) i / (double) NSHADES;

	for (i = 0; i < XPTS; i++) {
	   for (j = 0; j < YPTS; j++) {
	      // Replacement for mypltr of x16c.c
	      x = tr[0] * i + tr[1] * j + tr[2];
	      y = tr[3] * i + tr[4] * j + tr[5];

	      argx = x * Math.PI/2;
	      argy = y * Math.PI/2;
	      distort = 0.4;

	      // Note xg0 ==> yg1 are one-dimensional because of arrangement of
	      //zeros in the final tr definition above.  However, for now
	      //we are using raw interface here so must nominally treat them
	      //as two-dimensional.
	      xg0[i][j] = x;
	      yg0[i][j] = y;

	      xg1[i][j] = x + distort * Math.cos(argx);
	      yg1[i][j] = y - distort * Math.cos(argy);

	      xg2[i][j] = x + distort * Math.cos(argx) * Math.cos(argy);
	      yg2[i][j] = y - distort * Math.cos(argx) * Math.cos(argy);
	   }
	}

	// Plot using identity transform

	pls.adv(0);
	pls.vpor(0.1, 0.9, 0.1, 0.9);
	pls.wind(-1.0, 1.0, -1.0, 1.0);

	pls.psty(0);

	pls.shades(z, -1., 1., -1., 1., 
		 shedge, fill_width,
		 cont_color, cont_width,
		 1, xg0, yg0);

	pls.col0(1);
	pls.box("bcnst", 0.0, 0, "bcnstv", 0.0, 0);
	pls.col0(2);
	pls.lab("distance", "altitude", "Bogon density");

	// Plot using 1d coordinate transform
    
	pls.adv(0);
	pls.vpor(0.1, 0.9, 0.1, 0.9);
	pls.wind(-1.0, 1.0, -1.0, 1.0);

	pls.psty(0);

	pls.shades(z, -1., 1., -1., 1., 
	     shedge, fill_width,
	     cont_color, cont_width,
	     1, xg1, yg1);

	pls.col0(1);
	pls.box("bcnst", 0.0, 0, "bcnstv", 0.0, 0);
	pls.col0(2);

	pls.lab("distance", "altitude", "Bogon density");

	// Plot using 2d coordinate transform

	pls.adv(0);
	pls.vpor(0.1, 0.9, 0.1, 0.9);
	pls.wind(-1.0, 1.0, -1.0, 1.0);

	pls.psty(0);

	pls.shades(z, -1., 1., -1., 1., 
		 shedge, fill_width,
		 cont_color, cont_width,
		 0, xg2, yg2);

	pls.col0(1);
	pls.box("bcnst", 0.0, 0, "bcnstv", 0.0, 0);
	pls.col0(2);
	pls.cont(w, 1, XPTS, 1, YPTS, clevel, xg2, yg2);

	pls.lab("distance", "altitude", "Bogon density, with streamlines");

	// Plot using 2d coordinate transform

	pls.adv(0);
	pls.vpor(0.1, 0.9, 0.1, 0.9);
	pls.wind(-1.0, 1.0, -1.0, 1.0);

	pls.psty(0);

	pls.shades(z, -1., 1., -1., 1., 
		 shedge, fill_width,
		 2, 3,
		 0, xg2, yg2);

	pls.col0(1);
	pls.box("bcnst", 0.0, 0, "bcnstv", 0.0, 0);
	pls.col0(2);

	pls.lab("distance", "altitude", "Bogon density");

	// Note this exclusion API will probably change so don't bother
	// with it for x16.java example.


	//Example with polar coordinates.

	pls.adv(0);
	pls.vpor( .1, .9, .1, .9 );
	pls.wind( -1., 1., -1., 1. );

	pls.psty(0);

	// Build new coordinate matrices.
    
	for (i = 0; i < XPTS; i++) {
	   r = ((double) i)/ (XPTS-1);
	   for (j = 0; j < YPTS; j++) {
	      t = (2.*Math.PI/(YPTS-1.))*j;
	      xg2[i][j] = r*Math.cos(t);
	      yg2[i][j] = r*Math.sin(t);
	      z[i][j] = Math.exp(-r*r)*Math.cos(5.*Math.PI*r)*Math.cos(5.*t);
	   }
	}

	//Need a new shedge to go along with the new data set.

	f2mnmx( z, XPTS, YPTS );
	zmin = fmin;
	zmax = fmax;

	for (i = 0; i < NSHADES+1; i++)
	  shedge[i] = zmin + (zmax - zmin) * (double) i / (double) NSHADES;

	pls.shades(z, -1., 1., -1., 1., 
		 shedge, fill_width,
		 cont_color, cont_width,
		 0, xg2, yg2);

// Now we can draw the perimeter.  (If do before, shade stuff may overlap.)
	for (i = 0; i < PERIMETERPTS; i++) {
	   t = (2.*Math.PI/(PERIMETERPTS-1))*(double)i;
	   px[i] = Math.cos(t);
	   py[i] = Math.sin(t);
	}
	pls.col0(1);
	pls.line(px, py);
                  
	// And label the plot.

	pls.col0(2);
	pls.lab( "", "",  "Tokamak Bogon Instability" );

	// Clean up 

	pls.end();

     }
}

//---------------------------------------------------------------------------//
//                              End of x16.java
//---------------------------------------------------------------------------//

